### DISCUSSION TITLE
The problem of opencv path in xf_headers.hpp file.

### DISCUSSION THREAD

**Discussion by ecoyukino** [bol-edu/HLS-SOC-Discussions#43] [General]
Hi, this is 陳揚哲 and 周子翔. 
We are conducting Vision isppipeline_hdr topic in Lab-C. 
There's a problem of opencv directory.
In testbench, we need to include the xf_headers.hpp file, while the original code include the incorrect directory of opencv module.
Therefore, we modify the xf_headers.hpp file to the absolute directory.
We change the code from：

### #include "opencv2/video/video.hpp"

to：

### #include "/usr/local/include/opencv4/opencv2/video/video.hpp"

However, it turns out the .hpp file above also include the module with the incorrect directory.
The error message is shown below：

### /usr/local/include/opencv4/opencv2/opencv.hpp:48:10: fatal error: opencv2/opencv_modules.hpp: No such file or directory  #include "opencv2/opencv_modules.hpp"

Furthermore, these .hpp file is under /usr folder which can't be modified since we didn't have the permission.
We wonder that if we can get the permission to modify the file under the /usr folder, or is there any method to deal with this error.

**Comment by kevinjantw** [comment]
Hi, which u50_? server do you use?

**Reply by ecoyukino** [reply]
hi, we use u50_01 and 01.LcglIk.

**Comment by Charlee0207** [comment]
Hi, I am working on a special topic that uses OpenCV, and I have set up some compile options in Vitis. Here's how I do it:

---
In the Explorer panel,  right-click ``<ProjectName>.prj`` and select **Properties**. 
    Add the following includes and libraries in the **Properties**:
    
   * **C/C++ Build > Settings > Tool Setting > GCC Host Compiler(x86_64) > Includes**
     Add the path ``<path-to-opencv-include-folder>`` to **Include paths(-l)**.
   
   * **C/C++ Build > Settings > Tool Setting > GCC Host Linker(x86_64) > Libraries**
  
     Add the path ``<path-to-opencv-lib-folder>`` to **Library search path(-L)**.
     
     Add the following libraries:
       ``opencv_videoio`` ``opencv_imgcodecs`` ``opencv_core`` ``opencv_imgproc`` ``opencv_features2d`` ``opencv_flann`` ``opencv_video`` ``opencv_calib3d`` ``opencv_highgui``
       to **Libraries(-l)**.
---
On your environment, ``<path-to-opencv-include-folder>`` and ``<path-to-opencv-lib-folder>`` are likely to be ``/usr/local/include/opencv4`` and ``/usr/local/lib``, respectively.

If you are working on Vitis_hls, you may need to fill the compile options in CFLAGS and CSIMFLAGS.

**Reply by ZheChen-Bill** [reply]
Hi, I am another student is conducting this project.
I modified the setting as you told, but I still got the same error message.
I think it's because the relative path can't be detected correctly, while absolute path can.
just like we modify the "xf_headers.hpp" file
from:

**#include "opencv2/opencv.hpp"**

to:

**#include "/usr/local/include/opencv4/opencv2/opencv.hpp"**

And we could include the "opencv.hpp" file correctly.
Thus, we thought if we change all the relative path to absolute path may solve our problem.
Or, is there any other reason causing the relative path can't be detected? How can we solve this problem?

**Reply by kevinjantw** [reply]
Hi @ZheChen-Bill @ecoyukino ,
Please help to provide
(1) github link of original Xilinx vision case
(2) which files do you change for opencv lib path? (e.g. 01.LcglIk/Vitis-Tutorials/Hardware_Acceleration/Design_Tutorials/03-rtl_stream_kernel_integration/../xxx)
(3) log file of all executed linux command seqences + errors (can reproduce your errors)

It is more likely a project setting issue not a permission issue.

Kevin Jan

**Reply by ZheChen-Bill** [reply]
Hi, @kevinjantw 
(1) This is the github link of original Xilinx vision case
https://github.com/Xilinx/Vitis_Libraries/tree/master/vision
(2) We are testing the L1 function of gaincontrol can be implemented 
We modify the 01.LcglIk/Vitis_Libraries/vision/L1/examples/gaincontrol/Makefile 
![1682047139179](https://user-images.githubusercontent.com/88698677/233533039-265c562b-1241-499e-b835-73953e406462.jpg)
Type the code below in command line:
**make run**
After generate the required file for vitis_hls
Type the code below in command line to build the gaincontrol.prj:
**vitis_hls -f run_hls.tcl**
After the project is built, we use vitis_hls to open the project.
**vitis_hls -p gaincontrol.prj/**
Then modify the xf_headers.hpp in the path /01.LcglIk/Vitis_Libraries/vision/L1/include/common/
![1682047825594](https://user-images.githubusercontent.com/88698677/233534463-f48729ee-f4e9-459e-9821-b9d85d975c06.jpg)
(3)This is our error logs
![圖片](https://user-images.githubusercontent.com/88698677/233534585-5468c17b-0f79-4c0a-b44e-e0e2ccc41764.png)

**Reply by kevinjantw** [reply]
Hi @ZheChen-Bill @ecoyukino,
The opencv lib setting issue is resolved. Change in the 01.LcglIk/Vitis_Libraries/vision/L1/examples/gaincontrol/Makefile (you don't need to change xf_headers.hpp).
`OPENCV_LIB = /usr/lib/x86_64-linux-gnu/`
`OPENCV_INCLUDE = /usr/include/opencv4/`

Currently, you need to resolve the new coming undefined references.
(Notice: Offical mentioned that `Valid installation of Vitis™ 2022.2 or later version`; We currently support to Vitis™ 2022.1)

https://github.com/bol-edu/HLS-Discussions#vitis-opencv-lab-setup

Kevin Jan

```console
INFO: [SIM 2] *************** CSIM start ***************
INFO: [SIM 4] CSIM will launch GCC as the compiler.
   Compiling ../../../../xf_gaincontrol_tb.cpp in debug mode
   Compiling ../../../../xf_gaincontrol_accel.cpp in debug mode
   Generating csim.exe
In file included from /home/hls01/Vitis_Libraries/vision/L1/include/common/xf_structs.hpp:27,
                 from /home/hls01/Vitis_Libraries/vision/L1/include/common/xf_common.hpp:20,
                 from /home/hls01/Vitis_Libraries/vision/L1/include/common/xf_sw_utils.hpp:20,
                 from /home/hls01/Vitis_Libraries/vision/L1/include/common/xf_headers.hpp:28,
                 from ../../../../xf_gaincontrol_tb.cpp:17:
/opt/Xilinx/Vitis_HLS/2022.1/include/ap_axi_sdata.h:97:45: warning: variable templates only available with -std=c++14 or -std=gnu++14
 template <typename T> constexpr std::size_t bitwidth = sizeof(T) * CHAR_BIT;
                                             ^~~~~~~~
/opt/Xilinx/Vitis_HLS/2022.1/include/ap_axi_sdata.h:99:48: warning: variable templates only available with -std=c++14 or -std=gnu++14
 template <std::size_t W> constexpr std::size_t bitwidth<ap_int<W>> = W;
                                                ^~~~~~~~~~~~~~~~~~~
/opt/Xilinx/Vitis_HLS/2022.1/include/ap_axi_sdata.h:100:48: warning: variable templates only available with -std=c++14 or -std=gnu++14
 template <std::size_t W> constexpr std::size_t bitwidth<ap_uint<W>> = W;
                                                ^~~~~~~~~~~~~~~~~~~~
/opt/Xilinx/Vitis_HLS/2022.1/include/ap_axi_sdata.h:102:23: warning: variable templates only available with -std=c++14 or -std=gnu++14
 constexpr std::size_t bitwidth<ap_fixed<_AP_W, _AP_I, _AP_Q, _AP_O, _AP_N>> = _AP_W;
                       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/opt/Xilinx/Vitis_HLS/2022.1/include/ap_axi_sdata.h:104:23: warning: variable templates only available with -std=c++14 or -std=gnu++14
 constexpr std::size_t bitwidth<ap_ufixed<_AP_W, _AP_I, _AP_Q, _AP_O, _AP_N>> = _AP_W;
                       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/opt/Xilinx/Vitis_HLS/2022.1/include/ap_axi_sdata.h:107:23: warning: variable templates only available with -std=c++14 or -std=gnu++14
 constexpr std::size_t bytewidth = (bitwidth<T> + CHAR_BIT - 1) / CHAR_BIT;
                       ^~~~~~~~~
In file included from /home/hls01/Vitis_Libraries/vision/L1/include/common/xf_structs.hpp:27,
                 from /home/hls01/Vitis_Libraries/vision/L1/include/common/xf_common.hpp:20,
                 from ../../../../xf_gaincontrol_config.h:20,
                 from ../../../../xf_gaincontrol_accel.cpp:17:
/opt/Xilinx/Vitis_HLS/2022.1/include/ap_axi_sdata.h:97:45: warning: variable templates only available with -std=c++14 or -std=gnu++14
 template <typename T> constexpr std::size_t bitwidth = sizeof(T) * CHAR_BIT;
                                             ^~~~~~~~
/opt/Xilinx/Vitis_HLS/2022.1/include/ap_axi_sdata.h:99:48: warning: variable templates only available with -std=c++14 or -std=gnu++14
 template <std::size_t W> constexpr std::size_t bitwidth<ap_int<W>> = W;
                                                ^~~~~~~~~~~~~~~~~~~
/opt/Xilinx/Vitis_HLS/2022.1/include/ap_axi_sdata.h:100:48: warning: variable templates only available with -std=c++14 or -std=gnu++14
 template <std::size_t W> constexpr std::size_t bitwidth<ap_uint<W>> = W;
                                                ^~~~~~~~~~~~~~~~~~~~
/opt/Xilinx/Vitis_HLS/2022.1/include/ap_axi_sdata.h:102:23: warning: variable templates only available with -std=c++14 or -std=gnu++14
 constexpr std::size_t bitwidth<ap_fixed<_AP_W, _AP_I, _AP_Q, _AP_O, _AP_N>> = _AP_W;
                       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/opt/Xilinx/Vitis_HLS/2022.1/include/ap_axi_sdata.h:104:23: warning: variable templates only available with -std=c++14 or -std=gnu++14
 constexpr std::size_t bitwidth<ap_ufixed<_AP_W, _AP_I, _AP_Q, _AP_O, _AP_N>> = _AP_W;
                       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/opt/Xilinx/Vitis_HLS/2022.1/include/ap_axi_sdata.h:107:23: warning: variable templates only available with -std=c++14 or -std=gnu++14
 constexpr std::size_t bytewidth = (bitwidth<T> + CHAR_BIT - 1) / CHAR_BIT;
                       ^~~~~~~~~
/opt/Xilinx/Vivado/2022.1/tps/lnx64/binutils-2.37/bin/ld: obj/xf_gaincontrol_tb.o: in function `main':
/home/hls01/Vitis_Libraries/vision/L1/examples/gaincontrol/gaincontrol.prj/sol1/csim/build/../../../../xf_gaincontrol_tb.cpp:78: undefined reference to `cv::imread(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, int)'
/opt/Xilinx/Vivado/2022.1/tps/lnx64/binutils-2.37/bin/ld: /home/hls01/Vitis_Libraries/vision/L1/examples/gaincontrol/gaincontrol.prj/sol1/csim/build/../../../../xf_gaincontrol_tb.cpp:105: undefined reference to `cv::imwrite(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, cv::_InputArray const&, std::vector<int, std::allocator<int> > const&)'
/opt/Xilinx/Vivado/2022.1/tps/lnx64/binutils-2.37/bin/ld: /home/hls01/Vitis_Libraries/vision/L1/examples/gaincontrol/gaincontrol.prj/sol1/csim/build/../../../../xf_gaincontrol_tb.cpp:106: undefined reference to `cv::imwrite(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, cv::_InputArray const&, std::vector<int, std::allocator<int> > const&)'
/opt/Xilinx/Vivado/2022.1/tps/lnx64/binutils-2.37/bin/ld: /home/hls01/Vitis_Libraries/vision/L1/examples/gaincontrol/gaincontrol.prj/sol1/csim/build/../../../../xf_gaincontrol_tb.cpp:108: undefined reference to `cv::absdiff(cv::_InputArray const&, cv::_InputArray const&, cv::_OutputArray const&)'
/opt/Xilinx/Vivado/2022.1/tps/lnx64/binutils-2.37/bin/ld: /home/hls01/Vitis_Libraries/vision/L1/examples/gaincontrol/gaincontrol.prj/sol1/csim/build/../../../../xf_gaincontrol_tb.cpp:109: undefined reference to `cv::imwrite(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, cv::_InputArray const&, std::vector<int, std::allocator<int> > const&)'
/opt/Xilinx/Vivado/2022.1/tps/lnx64/binutils-2.37/bin/ld: obj/xf_gaincontrol_tb.o: in function `cv::Mat::Mat(cv::Mat const&)':
/usr/include/opencv4/./opencv2/core/mat.inl.hpp:538: undefined reference to `cv::Mat::copySize(cv::Mat const&)'
/opt/Xilinx/Vivado/2022.1/tps/lnx64/binutils-2.37/bin/ld: obj/xf_gaincontrol_tb.o: in function `cv::Mat::~Mat()':
/usr/include/opencv4/./opencv2/core/mat.inl.hpp:739: undefined reference to `cv::fastFree(void*)'
/opt/Xilinx/Vivado/2022.1/tps/lnx64/binutils-2.37/bin/ld: obj/xf_gaincontrol_tb.o: in function `cv::Mat::operator=(cv::Mat const&)':
/usr/include/opencv4/./opencv2/core/mat.inl.hpp:760: undefined reference to `cv::Mat::copySize(cv::Mat const&)'
/opt/Xilinx/Vivado/2022.1/tps/lnx64/binutils-2.37/bin/ld: obj/xf_gaincontrol_tb.o: in function `cv::Mat::clone() const':
/usr/include/opencv4/./opencv2/core/mat.inl.hpp:811: undefined reference to `cv::Mat::copyTo(cv::_OutputArray const&) const'
/opt/Xilinx/Vivado/2022.1/tps/lnx64/binutils-2.37/bin/ld: obj/xf_gaincontrol_tb.o: in function `cv::Mat::create(int, int, int)':
/usr/include/opencv4/./opencv2/core/mat.inl.hpp:831: undefined reference to `cv::Mat::create(int, int const*, int)'
/opt/Xilinx/Vivado/2022.1/tps/lnx64/binutils-2.37/bin/ld: obj/xf_gaincontrol_tb.o: in function `cv::Mat::release()':
/usr/include/opencv4/./opencv2/core/mat.inl.hpp:851: undefined reference to `cv::Mat::deallocate()'
/opt/Xilinx/Vivado/2022.1/tps/lnx64/binutils-2.37/bin/ld: obj/xf_gaincontrol_tb.o: in function `cv::Mat::operator=(cv::Mat&&)':
/usr/include/opencv4/./opencv2/core/mat.inl.hpp:1416: undefined reference to `cv::fastFree(void*)'
collect2: error: ld returned 1 exit status
make: *** [Makefile.rules:393: csim.exe] Error 1
ERR: [SIM 100] 'csim_design' failed: compilation error(s).
INFO: [SIM 3] *************** CSIM finish ***************
```

**Reply by ZheChen-Bill** [reply]
Hi @kevinjantw,
The issue is resolved. 
Thanks for your help.
